[BEAM-735] PAssertStreaming should make sure the assertion happened.

amitsela opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

amitsela commented:
The [ticket](https://issues.apache.org/jira/browse/BEAM-735) contains a clear explanation for why this is needed.

amitsela commented:
R: @kennknowles

kennknowles commented on runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/utils/PAssertStreaming.java:
```diff
@@ -18,58 +18,73 @@
-  /**
-   * Copied aggregator names from {@link org.apache.beam.sdk.testing.PAssert}.
-   */
-  static final String SUCCESS_COUNTER = "PAssertSuccess";
```
I'll review in more detail tomorrow, but can you use `PAssert` itself? It has such aggregators, https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java#L1054

amitsela replied:
I'll take a look, I remember I originally used it but when `PAssertStreaming` was introduced it became unused.

amitsela replied:
I should probably catch `AssertionError` and count success/fail instead of marker as `PAssert` does because an `Error` will not propagate and fail like an `Exception`.

amitsela commented:
I've used the `Aggregator` names from `PAssert`, but it doesn't seem like I can actually use any transformation there.
I've also added an explicit flag for graceful stops, for most of the time tests should stop immediately or they will keep executing a couple of more micro-batches (streaming) and may fail when they shouldn't.
Finally, I removed all calls to close() on batch tests because they weren't really needed - generally all `close()`, `waitUntil()`, etc. should be handled by BEAM-595.

amitsela commented:
Rebased on master to resolve the conflict we had due to rename of one of the streaming tests.

kennknowles commented on runners/spark/src/test/java/org/apache/beam/runners/spark/DeDupTest.java:
```diff
@@ -53,7 +53,6 @@ public void testRun() throws Exception {
 
-    EvaluationResult res = (EvaluationResult) p.run();
-    res.close();
+    p.run();
```
If I recall correctly, it is now `TestPipeline` that makes it so `run()` is blocking and terminating (for streaming cases, watching the counts of successes/failures and timing out). In the newly proposed APIs this may return immediately before anything is run.

kennknowles commented:
LGTM. Self-merge whenever you like.
